### PR TITLE
Properly fixes the fabricator problem

### LINKED ISSUE #18356: Exosuit Fabricator looses Ultra-laser uprgade after synching with research
(WEB REPORT BY: bigkuro REMOTE: 172.93.109.202:7777)
# Revision
6bdd0f18dc463ea9ad94b7f532251b42b37e08c6

# Description
The exosuit fabricator looses its ultra-high-precision micro-laser upgrade when it is synched up with research, making it fabricate at least three times slower.

# Steps to Reproduce
Unscrew an exosuit fabricator, upgrade it with a rapid-part exchanger with ultra micro lasers in it, rescrew the fabricator, unscrew & rescrew the robotics console or use the robotics console to synch up near devices, then head to the exosuit fabricator and press the synch up button. 

# What you Expected
The fabricator to fabricate faster after upgrading it and synching up with research to expand the list of available items.

# What Actually Happened
While it does synch up with research, the fabricator looses its ability to fabricate faster.

### LINKED ISSUE #20087: Exofabs
(WEB REPORT BY: iflashyou REMOTE: 172.93.109.202:7777)
Sync. the exofabs resets the material costs of all stuff to pre upgrade state

### PR DESCRIPTION
**Pull request by MadmanMartian**
Rips the bandaid off that is #20302 and properly sutures the wounds of #18356, #20087

closes #18356
closes #20087

 - Moves research's tech level stuff to a list by reference. Reference being the ID. All this so you don't need to loop through EVERY.FREAKING.RESEARCH.OBJECT to find the one you want.
 - Actually fixes the issue where syncing research and updating the internal components were using different math.

:cl:
 * bugfix: Fabricators now take research into account when refreshing internal components

### CODE REVIEW

File: code/modules/research/fabricators.dm [hunk 1]
```diff
@@ -423,30 +425,25 @@
...
-						ptime -= 2
-					diff = round(initial(time_coeff) - (initial(time_coeff)*(T.level+ptime))/25,0.1)
-					if(time_coeff!=diff)
-						time_coeff = diff
-						output+="Production routines updated.<br>"
+	var/diff
+	var/datum/tech/T = files.known_tech["materials"]
+	if(T && T.level > 1)
+		var/pmat = 0//Calculations to make up for the fact that these parts and tech modify the same thing
+		for(var/obj/item/weapon/stock_parts/micro_laser/Ml in component_parts)
```
Comment (Pathid) [inline code/modules/research/fabricators.dm]: This should have been swapped with manipulators in #13044 if I'm not mistaken.

File: code/modules/research/fabricators.dm [hunk 2]
```diff
@@ -423,30 +425,25 @@
...
+			pmat += Ml.rating - 1
+		diff = round(initial(resource_coeff) - (initial(resource_coeff)*(T.level+(pmat*3)))/25,0.01)
+		if(resource_coeff!=diff)
+			resource_coeff = diff
+			output+="Production efficiency increased.<br>"
+	T = files.known_tech["programming"]
+	if(T && T.level > 1)
+		var/ptime = 0
+		for(var/obj/item/weapon/stock_parts/manipulator/Ma in component_parts)
```
Comment (Pathid) [inline code/modules/research/fabricators.dm]: See above.

File: code/modules/research/fabricators.dm [hunk 3]
```diff
@@ -462,21 +459,22 @@
...
+			var/datum/tech/T = console.files.known_tech[ID]
 			if(T)
 				files.AddTech2Known(T)
 		for(var/datum/design/D in console.files.known_designs)
 			if(D)
 				files.AddDesign2Known(D)
 		files.RefreshResearch()
 		var/i = src.convert_designs()
-		//var/tech_output = update_tech() //apparently this has never worked
+		var/tech_output = update_tech() //apparently this has never worked
```
Comment (SonixApache) [inline code/modules/research/fabricators.dm]: you can remove this comment

File: code/modules/research/fabricators.dm (hunk 1 above)
Comment (MadmanMartian) [reply code/modules/research/fabricators.dm]: Looks like that's the PR that caused the discrepency

**Pull request merged**